File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.c
int lengthOfLongestSubstring(char * s) {
    int n = strlen(s);
    int maxLen = 0;
    int start = 0;
    int index[128]; // Assuming ASCII characters

    for (int i = 0; i < 128; i++) {
        index[i] = -1; // Initialize index of characters as -1
    }

    for (int end = 0; end < n; end++) {
        char currentChar = s[end];
        if (index[(int)currentChar] >= start) {
            start = index[(int)currentChar] + 1;
        }
        index[(int)currentChar] = end;
        maxLen = (maxLen > (end - start + 1)) ? maxLen : (end - start + 1);
    }

    return maxLen;
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxLen = 0;
        int start = 0;
        Map<Character, Integer> indexMap = new HashMap<>();

        for (int end = 0; end < n; end++) {
            char currentChar = s.charAt(end);
            if (indexMap.containsKey(currentChar) && indexMap.get(currentChar) >= start) {
                start = indexMap.get(currentChar) + 1;
            }
            indexMap.put(currentChar, end);
            maxLen = Math.max(maxLen, end - start + 1);
        }

        return maxLen;
    }
}